pkg/process/finders/vm/finder.go

```diff
@@ -0,0 +1,318 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package vm
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	v3 "skywalking.apache.org/repo/goapi/collect/ebpf/profiling/process/v3"
+
+	"github.com/shirou/gopsutil/process"
+
+	"github.com/apache/skywalking-rover/pkg/logger"
+	"github.com/apache/skywalking-rover/pkg/process/api"
+	"github.com/apache/skywalking-rover/pkg/process/finders/base"
+	"github.com/apache/skywalking-rover/pkg/tools"
+)
+
+var log = logger.GetLogger("process", "finder", "vm")
+
+type ProcessFinder struct {
+	conf *Config
+
+	manager   base.ProcessManager
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+
+	period time.Duration
+}
+
+func (p *ProcessFinder) Init(ctx context.Context, conf base.FinderBaseConfig, manager base.ProcessManager) error {
+	if err := validateConfig(conf.(*Config)); err != nil {
+		return err
+	}
+
+	p.conf = conf.(*Config)
+	p.manager = manager
+	p.ctx, p.cancelCtx = context.WithCancel(ctx)
+
+	period, err := time.ParseDuration(p.conf.Period)
+	if err != nil {
+		return err
+	}
+	p.period = period
+	return nil
+}
+
+func (p *ProcessFinder) Start() {
+	go p.startWatch()
+}
+
+func (p *ProcessFinder) Stop() error {
+	p.cancelCtx()
+	return nil
+}
+
+func (p *ProcessFinder) DetectType() api.ProcessDetectType {
+	return api.VM
+}
+
+func (p *ProcessFinder) ValidateProcessIsSame(p1, p2 base.DetectedProcess) bool {
+	vm1 := p1.(*Process)
+	vm2 := p2.(*Process)
+	return p1.Pid() == p2.Pid() && vm1.cmd == vm2.cmd && p1.Entity().SameWith(p2.Entity())
+}
+
+func (p *ProcessFinder) BuildEBPFProcess(ctx *base.BuildEBPFProcessContext, ps base.DetectedProcess) *v3.EBPFProcessProperties {
+	hostProcess := &v3.EBPFHostProcessMetadata{}
+	hostProcess.HostIP = ctx.HostIP
+	hostProcess.Pid = ps.Pid()
+	hostProcess.Cmd = ps.(*Process).cmd
+	hostProcess.Entity = &v3.EBPFProcessEntityMetadata{
```

Comment:
Do we have any mechanism to verify whether this entity is duplicated? Or should we check on the server side?

Reply:
Yes, I have a mechanism to verify their entity is the same. Each VM process has [an identity](https://github.com/apache/skywalking-rover/pull/5/files#diff-ed74b624b51883d381869a88732798903eb04b05ec50f7b2435a25eb7cfc0a6eR64-R67), If the same identity could have multiple matched processes, then I would [only use the first one and log them](https://github.com/apache/skywalking-rover/pull/5/files#diff-9ac007c61d6cca27e5e4c59578707fa481fdb52a26f9fa26f8f841862b027607R210-R236).